### PR TITLE
feat: support vllm

### PR DESCRIPTION
**Pull request by jemmyshin**
integrate with vllm to support efficient kv cache management and tensor parallelism

First, implement this in `generate` endpoint to run benchmark

### CODE REVIEW

**Comment by numb3r3** [comment]
@jemmyshin Please explain the benefits caused by vllm integration. It is better to exploit the improvements in terms of `page attention` and `tensor_parallism` respectively.

File: open_gpt/cli/commands/deploy.py [hunk 1]
```diff
@@ -30,6 +30,12 @@ class DeployCommand(Command):
             flag=False,
             default='balanced',
         ),
+        option(
+            'backend',
+            None,
+            'The backend of the model.',
+            default='hf',
+        ),
```
Comment (numb3r3) [inline open_gpt/cli/commands/deploy.py:38]: I would suggest opengpt could decide which backend to use in the runtime. The principle of opengpt is to obtain the best pratice of llm serving out of the box.

File: open_gpt/models/generation.py [hunk 2]
```diff
@@ -365,27 +366,62 @@ def generate(self, prompt: str, max_length: int = MAX_LENGTH, **kwargs):
...
+                                             temperature=kwargs.pop('temperature', 1.0),
+                                             top_p=kwargs.pop('top_p', 0.9),
+                                             top_k=kwargs.pop('top_k', 1),
+                                             use_beam_search=kwargs.pop('do_sample', False),
+                                             stop=stop_str,
+                                             ignore_eos=True,
+                                             max_tokens=max_new_tokens,
+                                             logprobs=kwargs.pop('logprobs', None),
+                                             )
+            output = self.model.generate(prompt, sampling_params=sampling_params, use_tqdm=False)[0]
```
Comment (numb3r3) [inline open_gpt/models/generation.py:412]: It is better to place the generation of vllm in an individual python module `vllm_generation.py`.

File: open_gpt/models/modeling.py [hunk 3]
```diff
@@ -34,21 +38,25 @@ def __init__(
...
+        self._backend = backend
         self._precision = precision
         self._dtype, self._device = auto_dtype_and_device(precision, device)
 
         self._device_map = device_map or get_device_map(self._device)
 
         self._eval_mode = eval_mode
 
+        self._tensor_parallel_size = tensor_parallel_size
+        self._pipeline_parallel_size = pipeline_parallel_size
```
Comment (numb3r3) [inline open_gpt/models/modeling.py:50]: All of these hyperparameters should be advanced configurations and would be better to be determined automatically without manual settings.

File: open_gpt/models/modeling.py [hunk 4]
```diff
@@ -61,18 +69,35 @@ def load_model_and_transforms(
...
+            from vllm import LLM
+
+            self.model = LLM(model=model_name_or_path,
+                             tokenizer=tokenizer_name_or_path,
+                             tokenizer_mode='slow',
+                             tensor_parallel_size=mself._tensor_parallel_size or torch.cuda.device_count(),
+                             # Pipeline parallelism is not supported yet.
+                             pipeline_parallel_size=1,
+                             trust_remote_code=True)
+            self.tokenizer = self.model.get_tokenizer()
```
Comment (numb3r3) [inline open_gpt/models/modeling.py:100]: I guess it cannot work with LoRA finetuning as vllm has its own model implementation.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: open_gpt/models/modeling.py [hunk 5]
```diff
@@ -34,21 +38,25 @@ def __init__(
...
+
         self.load_model_and_transforms(
             model_name_or_path,
             adapter_name_or_path=adapter_name_or_path,
             tokenizer_name_or_path=tokenizer_name_or_path,
         )
 
         # turn the eval mode off `eval_mode=False` in training
-        if self._eval_mode:
+        if self._eval_mode and self._backend == 'hf':
```
Comment (numb3r3) [inline open_gpt/models/modeling.py:59]: why `self._backen==hf` means `eval` mode ?

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by jemmyshin** [comment]
I will address these comments after investigating lightllm since we haven't decided which one to use in the end.

File: open_gpt/cli/commands/deploy.py (hunk 1 above)
Comment (jemmyshin) [reply open_gpt/cli/commands/deploy.py]: `vllm` doesn't support streaming output, so if this the default option, it will break the whole user experience.

Comment (numb3r3) [reply open_gpt/cli/commands/deploy.py]: @jemmyshin If this is the case, we should hold the vllm integration.

Comment (numb3r3) [reply open_gpt/cli/commands/deploy.py]: vllm is not the good backend if it does not have stream supporting.

**Pull request closed without merging**